// angular-routing/esm2015/lib/router.module.js
var RoutingModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { LocationStrategy, PathLocationStrategy, CommonModule, } from '@angular/common';
import { RouterComponent } from './router.component';
import { RouteComponent } from './route.component';
import { RouteComponentTemplate } from './route-component.directive';
import { LinkActive } from './link-active.directive';
import { LinkTo } from './link-to.directive';
import { UrlParser } from './url-parser';
import { QueryParams } from './route-params.service';
import { Router } from './router.service';
export const components = [
    RouterComponent,
    RouteComponent,
    LinkActive,
    LinkTo,
    RouteComponentTemplate,
];
export function getQueryParams(router) {
    return router.queryParams$;
}
let RoutingModule = RoutingModule_1 = class RoutingModule {
    static forRoot() {
        return {
            ngModule: RoutingModule_1,
            providers: [
                UrlParser,
                { provide: LocationStrategy, useClass: PathLocationStrategy },
                { provide: QueryParams, deps: [Router], useFactory: getQueryParams },
            ],
        };
    }
};
RoutingModule = RoutingModule_1 = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [components],
        exports: [components],
    })
], RoutingModule);
export { RoutingModule };
